### ISSUE TITLE
Error loading preference dialog

### ISSUE THREAD

**Issue by MoustafaC** [micheleg/dash-to-dock#480]
I've seen this error pop up on Ubuntu 17.04 while loading the preferences dialog:
```
Error: Requiring Clutter, version none: Typelib file for namespace 'Clutter' (any version) not found

Stack trace:
  @/home/steam/.local/share/gnome-shell/extensions/dash-to-dock@micxgx.gmail.com/convenience.js:8:7
  @/home/steam/.local/share/gnome-shell/extensions/dash-to-dock@micxgx.gmail.com/prefs.js:17:7
  Application<._getExtensionPrefsModule@resource:///org/gnome/shell/extensionPrefs/main.js:74:13
  wrapper@resource:///org/gnome/gjs/modules/lang.js:178:22
  Application<._selectExtension@resource:///org/gnome/shell/extensionPrefs/main.js:89:31
  wrapper@resource:///org/gnome/gjs/modules/lang.js:178:22
  Application<._onCommandLine@resource:///org/gnome/shell/extensionPrefs/main.js:243:17
  wrapper@resource:///org/gnome/gjs/modules/lang.js:178:22
  main@resource:///org/gnome/shell/extensionPrefs/main.js:399:5
  @<main>:1:49
```

This has happened even after removing and re-installing the extension.

Sysinfo:
```
Shell Version: 3.24
Extension version: v57
Linux Distribution and version: Ubuntu 17.04
Multiple monitor: Yes
Settings:
[/]
autohide-in-fullscreen=false
dock-fixed=false
intellihide-mode='FOCUS_APPLICATION_WINDOWS'
click-action='minimize'
autohide=true
apply-custom-theme=false
require-pressure-to-show=true
dock-position='BOTTOM'
custom-theme-shrink=true
preferred-monitor=0
dash-max-icon-size=48
opaque-background=true
shortcut=['<Super>q']
```
Looking through various log files gave no additional information as to why this issue is occurring.
Here's the output of dpkg -l *clutter*
```
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                       Version       Architecture Description
+++-==========================-=============-============-===========================================================
un  gstreamer1.0-clutter       <none>        <none>       (no description available)
ii  gstreamer1.0-clutter-3.0   3.0.24-1      amd64        Clutter PLugin for GStreamer 1.0
un  ibus-clutter               <none>        <none>       (no description available)
ii  libclutter-1.0-0:amd64     1.26.0+dfsg-3 amd64        Open GL based interactive canvas library
ii  libclutter-1.0-common      1.26.0+dfsg-3 all          Open GL based interactive canvas library (common files)
un  libclutter-gst-2.0-0       <none>        <none>       (no description available)
ii  libclutter-gst-3.0-0:amd64 3.0.24-1      amd64        Open GL based interactive canvas library GStreamer elements
ii  libclutter-gtk-1.0-0:amd64 1.8.2-2       amd64        Open GL based interactive canvas library GTK+ widget

```

**Comment by franglais125** [comment]
Can you try installing `gir1.2-clutter-1.0` ?

My guess is you installed `gnome-shell` directly, instead of the meta-package?

**Comment by MoustafaC** [comment]
That corrected it. Which is weird, since i had been using Dash-to-Dock for some time, but it only broke after I updated to 17.04. Other users on the GNOME extensions page have had a similar issue, which is why I opened the bug.

Consider this fixed.

**Comment by franglais125** [comment]
It's actually a duplicate of #398, but for some reason I've seen this happen with 17.04 (due to reports on this extension and others as well) very often.

I don't know what the problem is with the 17.04 packages, but some dependency (or upgrade path) seems to be off.

Glad it's fixed for you.

**Comment by micheleg** [comment]
Let's keep this bug open, as I have received notification of this problem multiple times in the last couple of weeks. I've also made a [post](https://micheleg.github.io/dash-to-dock/news/2017/04/23/ubuntu-settings-error-clutter.html) on the website for people to find out the issue.

**Comment by MacLotsen** [comment]
Also on Arch Linux it's not working. I have the following packages installed concerning clutter:
```
$ pacman -Qs clutter
local/clutter 1.26.2-1
    A toolkit for creating fast, portable, compelling dynamic UIs
local/clutter-gst 3.0.24-1
    GStreamer bindings for clutter
local/clutter-gtk 1.8.2-1
    Clutter integration with GTK+
```
A package like gir*clutter does not exist. Gir comes with the package clutter-gtk though. Does someone have any leads for the package I'm missing?

**Comment by dasnoopy** [comment]
Hi,all... I recently noticed same issue as described  also on  ArchLinux and Gnome 3.24.2

Issue **occured** (at least for me) only if I manually install  then extension using latest git version
Issue **doesn't** occured if I use the version available on the gnome shell extension website 

On archlinux,  clutter packages  available in extra repository are these: 
```
clutter 1.26.2-1
clutter-gst 3.0.24-1
clutter-gtk 1.8.2-1
```
and there's any packages related to 
regards

**Comment by franglais125** [comment]
@dasnoopy can you try installing the 3 packages you listed? At least on my system they were installed as dependencies.

PS: I don't use Arch, I'm judging by version numbers, comparing to Debian testing.
```
gir1.2-clutter-1.0        1.26.0+dfsg-3
gir1.2-clutter-gst-3.0    3.0.24-1
gir1.2-gtkclutter-1.0     1.8.2-2
```

**Comment by dasnoopy** [comment]
.. all packages related clutter are already installed in the system. Then I checked the clutter* packages content since there's any gir*-clutter packages in repositories,   and  gir files are already  included in the relative package. So this is way on ARCH there's no separate gir* packages..

I tried to isolate the issue (at least on Arch)  and I discovered that the "cultprit" is this recent commit https://github.com/micheleg/dash-to-dock/commit/ab2c13182e63023beba752d13c8e60e07ef3d1b6 : reverting the change, preferences windows appears and everything works fine....

Regards
Andrea

**Comment by micheleg** [comment]
Thanks Andrea for the additional information. This is very weird. As mentioned in the commit message, that commit was specifically solving a similar issue for some other user. Can you please confirm that you can repeoduce the behaviour with/without that commit, i.e. that it wasn't accidental that it started working. Also, can you confirm which is the error message that you get when the preferences panel cannot start?

**Comment by franglais125** [comment]
@micheleg Out of curiosity I installed your master branch on Ubuntu 17.04 (so Gnome 3.24), and I can reproduce the issue. Here is an screenshot:
![screenshot from 2017-05-28 13-46-47](https://cloud.githubusercontent.com/assets/19195975/26530898/6fa2cacc-43ac-11e7-9ff5-9768e06b0a95.png)

I locally reverted commit [ab2c131], restarted the session and got things working fine again.

No errors are reported on the journal.

**Comment by franglais125** [comment]
I was thinking... `St` is only used in the `getPosition()` function. Maybe we don't see often the problem that [ab2c13182e63023beba752d13c8e60e07ef3d1b6] was trying to fix, as that particular piece of code is only executed for RTL languages.

Maybe we can move this function to another .js file?

**Comment by franglais125** [comment]
I can confirm the suspicion: when removing the RTL if condition, loading the extension fails on Gnome 3.24 if `St` is not imported.

I added a quick fix for this here: [ab2c13182e63023beba752d13c8e60e07ef3d1b6]. @dasnoopy can you try it out?

@micheleg I don't know how much of a fan of this you are though!

**Comment by micheleg** [comment]
I guess you meant to link to https://github.com/franglais125/dash-to-dock/commit/6390b689b8dcc67e006e8dfb8aea55a8ea827d80?
Good point that of RTL languages. So you are saying that that commit ab2c131 fix RTL but break LTR mode?

**Comment by franglais125** [comment]
Ah, yes, sorry, bad commit link!

[ab2c13182e63023beba752d13c8e60e07ef3d1b6] is necessary to load the extension for RTL languages, if not it fails with the error you included in the commit message.

However, on Gnome 3.24, it prevents the user from opening the settings dialog, and this is independent of RTL or not.

https://github.com/franglais125/dash-to-dock/commit/6390b689b8dcc67e006e8dfb8aea55a8ea827d80 takes care of both problems, as it declares `St` for `getPosition` to work properly, but at the same time it's not defined globally, so it doesn't interfere with settings loading.

I'd like to get confirmation that it fixes both problems. It did for me at least in my testing!

**Comment by micheleg** [comment]
However, I agree we should probably move that function somewhere else. It probably makes little sense to import St (the shell toolkit) outside of the shell. Indeed trying running with gjs a simple js script fails importing St. How it has been working till now with no error I don't know.

**Comment by micheleg** [comment]
Moving the function to something like utils.js might also solve the issue in ubuntu with missin Clutter, as we wouldn't need to import Clutter in the preferences context at all, therefore not requiring to install the Clutter gir files.

**Comment by franglais125** [comment]
I can take care of doing that if you want. Should we move the signal/injection handlers as well? Just to keep convenience.js rather minimal.

**Comment by micheleg** [comment]
Yes, I think it's a good idea. Convenience was originally imported from the upstream extensions, mainly to provide translation and settings support. It makes sense to keep it like that, and move our custom code somewhere else. I'm happy if you go on with this. utils.js seems to me a good source name, unless you can think of something better.

**Comment by franglais125** [comment]
I moved the functions/classes to utils.js.

Here is the branch: https://github.com/franglais125/dash-to-dock/tree/st_get_position

I'll give it some testing on 3.22 and 3.24 and report back.

**Comment by micheleg** [comment]
This was fixed in #509.
